### PR TITLE
feat(collision_detector): add autoware_collision_detector

### PR DESCRIPTION
**Pull request by go-sakayori**
## Description
This PR contains the initial implementation of the autoware_collision_detector.
Please refer to this [issue](https://github.com/autowarefoundation/autoware.universe/issues/9145) for further information.

## Related links

**Parent Issue:**
- https://github.com/autowarefoundation/autoware.universe/issues/9145

## How was this PR tested?
Colcon build is successful 

## Notes for reviewers
Launch files for control would be changed in a different PR

### Input

| Name                                           | Type                                                   | Description                                                        |
| ---------------------------------------------- | ------------------------------------------------------ | ------------------------------------------------------------------ |
| `/perception/obstacle_segmentation/pointcloud` | `sensor_msgs::msg::PointCloud2`                        | Pointcloud of obstacles which the ego-vehicle should stop or avoid |
| `/perception/object_recognition/objects`       | `autoware_perception_msgs::msg::PredictedObjects` | Dynamic objects                                                    |
| `/tf`                                          | `tf2_msgs::msg::TFMessage`                             | TF                                                                 |
| `/tf_static`                                   | `tf2_msgs::msg::TFMessage`                             | TF static                                                          |

### Output

| Name           | Type                                    | Description |
| -------------- | --------------------------------------- | ----------- |
| `/diagnostics` | `diagnostic_msgs::msg::DiagnosticArray` | Diagnostics |

## Parameters

| Name                         | Type                    | Description                                                                              | Default value                    |
| :--------------------------- | :---------------------- | :--------------------------------------------------------------------------------------- | :------------------------------- |
| `use_pointcloud`             | `bool`                  | Use pointcloud as obstacle check                                                         | `true`                           |
| `use_dynamic_object`         | `bool`                  | Use dynamic object as obstacle check                                                     | `true`                           |
| `collision_distance`         | `double`                | Distance threshold at which an object is considered a collision. [m]                     | 0.15                             |
| `nearby_filter_radius`       | `double`                | Distance range for filtering objects. Objects within this radius are considered. [m]     | 5.0                              |
| `keep_ignoring_time`         | `double`                | Time to keep filtering objects that first appeared in the vicinity [sec]                 | 10.0                             |
| `nearby_object_type_filters` | `object of bool values` | Specifies which object types to filter. Only objects with `true` value will be filtered. | `{unknown: true, others: false}` |

## Effects on system behavior

None.

### CODE REVIEW

File: control/autoware_collision_detector/src/node.cpp [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+    auto observed_it = std::find_if(
+      observed_objects_.begin(), observed_objects_.end(), [&object](const auto & observed_object) {
+        return observed_object.object_id == object.object_id;
+      });
+    const bool was_observed = (observed_it != observed_objects_.end());
+
+    if (was_observed) {
+      observed_it->timestamp = current_object_time;
+      // Add without filtering
+      filtered_objects.objects.push_back(object);
```
Comment (ambroise-arm) [inline control/autoware_collision_detector/src/node.cpp:277]: I am confused about what "filtering" means here. The comment says we are not filtering, but we are adding the object to a "filtered" list?

File: control/autoware_collision_detector/src/node.cpp [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+  bg::correct(object_polygon);
+
+  return object_polygon;
+}
+
+Polygon2d createObjPolygon(
+  const geometry_msgs::msg::Pose & pose, const geometry_msgs::msg::Vector3 & size)
+{
+  const double & length_m = size.x / 2.0;
+  const double & width_m = size.y / 2.0;
```
Comment (ambroise-arm) [inline control/autoware_collision_detector/src/node.cpp:78]: Just curious, why are we using references here?

Comment (go-sakayori) [reply control/autoware_collision_detector/src/node.cpp]: Thank you for your review. No reference is needed here, so I've fixed it in cfafc53a691189bb9c47f2ecbf0b4fa285485bbb

File: control/autoware_collision_detector/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+## Parameters
+
+| Name                         | Type                    | Description                                                                                     | Default value                    |
+| :--------------------------- | :---------------------- | :---------------------------------------------------------------------------------------------- | :------------------------------- |
+| `use_pointcloud`             | `bool`                  | Use pointcloud as obstacle check                                                                | `true`                           |
+| `use_dynamic_object`         | `bool`                  | Use dynamic object as obstacle check                                                            | `true`                           |
+| `collision_distance`         | `double`                | If objects exist in this distance, publish error diagnostics [m]                                | 0.15                             |
+| `nearby_filter_radius`       | `double`                | If objects appear in this distance with specified classification, publish error diagnostics [m] | 5.0                              |
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply control/autoware_collision_detector/README.md:70]: @ambroise-arm 
Thank you so much for the comment! Your comment is correct!
I will update with other mistake

File: control/autoware_collision_detector/src/node.cpp (hunk 1 above)
Comment (kyoichi-sugahara) [reply control/autoware_collision_detector/src/node.cpp:277]: @ambroise-arm 
Thank you for your comments. I realized that my use of "filter" was inappropriate. 
The term `filtered_objects` represents a list of objects that meet the specified conditions, but I mistakenly used `filtering` to describe the exclusion of objects, which caused confusion. Therefore, I plan to unify the terminology by explicitly indicating that `filtered_objects` is a list of objects meeting the conditions and avoid using `filter` to describe exclusion. If you agree, I will proceed with the changes based on this approach.

File: control/autoware_collision_detector/package.xml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>autoware_collision_detector</name>
+  <version>0.1.0</version>
+  <description>The collision_detector package</description>
+  <maintainer email="shinnosuke.hirakawa@tier4.jp">Shinnosuke Hirakawa</maintainer>
```
Comment (0x126) [inline control/autoware_collision_detector/package.xml:7]: Please remove me as a maintainer.

Comment (go-sakayori) [reply control/autoware_collision_detector/package.xml]: Fixed in 80d5e17fd9bd3bf78ee65b990c2a69e5d26a5b9f

File: control/autoware_collision_detector/src/node.cpp (hunk 1 above)
Comment (ambroise-arm) [reply control/autoware_collision_detector/src/node.cpp:277]: Sounds good

File: control/autoware_collision_detector/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (go-sakayori) [reply control/autoware_collision_detector/README.md:70]: @ambroise-arm
Would these definition be better?
| Name                         | Type                    | Description                                                                                     | Default value                    |
| :--------------------------- | :---------------------- | :---------------------------------------------------------------------------------------------- | :------------------------------- |
| `collision_distance`         | `double`                | Distance threshold at which an object is considered a collision. [m]                       | 0.15                             |
| `nearby_filter_radius`       | `double`                | Distance range for filtering objects. Objects within this radius are considered. [m]            | 5.0                              |

File: control/autoware_collision_detector/src/node.cpp (hunk 1 above)
Comment (go-sakayori) [reply control/autoware_collision_detector/src/node.cpp]: Definition changed in 9b89d41266d4f0ae620843e6794376e30d496032

File: control/autoware_collision_detector/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ambroise-arm) [reply control/autoware_collision_detector/README.md:70]: Yes!

[the remark this answers is not in the document]
Comment (go-sakayori) [reply control/autoware_collision_detector/README.md]: Changed the description in cd5d3c9d70fa9d94ec64791c926ac54304445ae7

**Review by kyoichi-sugahara — APPROVED** [review]
@go-sakayori 
Thank you so much for dealing with the all commented fix.

@ambroise-arm 
Thank you so much for attentive review!!

**Pull request merged**